[4.3.0] Add filter for the query for reserved stock

Chouby opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Background: Multilingual plugins such as WPML and Polylang have one product per language and maintain a synchronization of the stock across products translations. 

#25708 introduces a way to reserve stock for a product. To allow a perfect synchronization with the product translations, we should be able to reserve stock for a product and its translations at the same time.

I propose to add this new filter to allow multilingual plugins to add the product translations in the list of products to reserve stock.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

Commit:
Add filter for products to reserve stock

Commit:
Remove extra @since

mikejolley commented:
> #25708 introduces a way to reserve stock for a product. To allow a perfect synchronization with the product translations, we should be able to reserve stock for a product and its translations at the same time.

Just a question, but would it be less brittle if the filters were on read? So that for example, when checking stock you would map the localised ID to the original ID and use it's stock? Otherwise you are introducing potential sync issues and inserting way more rows that needed.

Chouby commented:
You are right, it would be better as this would not increase the DB write access.

In that case, I foresee that we would need to make [this query](https://github.com/woocommerce/woocommerce/blob/4.1.0-beta.2/src/Checkout/Helpers/ReserveStock.php#L163-L174) accept an array of product ids instead of one unique product id. Something like this:
```php
		return $wpdb->prepare(
			"
			SELECT COALESCE( SUM( stock_table.`stock_quantity` ), 0 ) FROM $wpdb->wc_reserved_stock stock_table
			LEFT JOIN $wpdb->posts posts ON stock_table.`order_id` = posts.ID
			WHERE posts.post_status IN ( 'wc-checkout-draft', 'wc-pending' )
			AND stock_table.`expires` > NOW()
			AND stock_table.`product_id` IN ( %s )
			AND stock_table.`order_id` != %d
			",
			implode( ',', $product_ids ),
			$exclude_order_id
		);
``` 

I foresee 2 ways for WooCommerce to accept this.
1. Filter the complete query with a `woocommerce_query_for_reserved_stock` filter with 3 parameters: the query itself, `$product_id`, and `$exclude_order_id`. 
2 Modify the query in WooCommerce core as proposed above to accept an array of product ids by default and add a filter `woocommerce_product_ids_for_reserved_stock` with 2 parameters: `array( $product_id )`, and `$exclude_order_id`. 

What would you prefer?

mikejolley commented:
Why an array? Wouldn't it be a 1:1 map?

Chouby commented:
Let say that a customer has reserved a product in English and another has reserved the same product in French (with a different product id) we need to sum the reserved stocks for the 2 products which are physically sharing the same stock.

mikejolley commented:
@Chouby isn't that backward? Remember the table contains the source product IDs - so in that case both products would map to the same ID, no sum required.

Chouby commented:
Currently, when a stock is reserved for a product in English, there is no way to also reserve the same stock for the translated product in French. 

I took your phunit test here: https://github.com/woocommerce/woocommerce/blob/0cce3c5c45af7c26aa8908dc7bf47dc7a42d501e/tests/unit-tests/checkout/checkout.php#L195-L206

And I modified it to:  

```php
protected function create_order_for_managed_inventory_product() {
	// Create a product in English.
	$en = WC_Helper_Product::create_simple_product();
	$en->set_manage_stock( true );
	$en->set_stock_quantity( 10 );
	$en->save();
	self::$data_store->set_language( $en->get_id(), 'en' );

	// Create a product in French.
	$fr = WC_Helper_Product::create_simple_product();
	$fr->set_manage_stock( true );
	$fr->set_stock_quantity( 10 );
	$fr->save();
	self::$data_store->set_language( $fr->get_id(), 'fr' );

	// Both products are translations of each other.
	self::$data_store->save_translations( array( 'en' => $en->get_id(), 'fr' => $fr->get_id() ) );

	new PLLWC_Stock();

	// Order the English product.
	WC()->cart->add_to_cart( $en->get_id(), 9 );
	$this->assertEquals( true, WC()->cart->check_cart_items() );

	$checkout = WC_Checkout::instance();
	$order_id = $checkout->create_order(
		array(
			'payment_method' => 'cod',
			'billing_email'  => 'a@b.com',
		)
	);

	// Assertions whether the order was created successfully.
	$this->assertNotWPError( $order_id );

	return array(
		wc_get_product( $en->get_id() ),
		wc_get_product( $fr->get_id() ),
		wc_get_order( $order_id ),
	);
}

public function test_create_order_when_out_of_stock() {
	list( $en, $fr, $order ) = $this->create_order_for_managed_inventory_product();

	$this->assertEquals( 9, $order->get_item_count() );
	$this->assertEquals( 'pending', $order->get_status() );

	$this->assertEquals( 9, wc_get_held_stock_quantity( $en ) ); // This assertion passes as in your test
	$this->assertEquals( 9, wc_get_held_stock_quantity( $fr ) ); // This test fails.

	WC()->cart->empty_cart();
	WC()->cart->add_to_cart( $en->get_stock_managed_by_id(), 2 ); // This assertion passes as in your test

	$this->assertEquals( false, WC()->cart->check_cart_items() );

	WC()->cart->empty_cart();
	WC()->cart->add_to_cart( $fr->get_stock_managed_by_id(), 2 );  // This test fails.

	$this->assertEquals( false, WC()->cart->check_cart_items() );
}
```

If I modify the query to make the sum for all the stock reserved  for all the translations of a product, then the two failing test would pass. I must make the sum, for all the translations of a product because I could have a reserved stock of 2 for the English product, 3 for the French product, 5 for the German products so in total, the reserved stock is 10.

claudiosanches commented:
@Chouby I agree with you, sounds better allow change that whole query since it's a custom behavior.

Commit:
add filter woocommerce_query_for_reserved_stock

Remove previosuly proposed filter woocommerce_reserve_stock_for_products

Chouby commented:
Now that #26395 re-introduces the feature for teh version 4.3.0, I update this PR.

@claudiosanches, based on your suggestion, I removed my previous filter proposal which wasn't the best solution and added a new filter `woocommerce_query_for_reserved_stock` to filter the whole query.

claudiosanches reviewed (approved):
Looks good, thanks!

claudiosanches committed:
Updated since tag

rrennick reviewed (approved):
Still looks good.
